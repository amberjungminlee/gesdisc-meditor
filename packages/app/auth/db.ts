import { getDb } from '../lib/connections'
import { makeSafeObjectIDs } from '../lib/mongodb'
import type { Db } from 'mongodb'
import type { Document } from '../documents/types'
import type { Model, ModelWithWorkflow } from '../models/types'
import type { UserContactInformation } from './types'

class UsersDb {
    #USERS_COLLECTION = 'Users'
    #ACCOUNTS_COLLECTION = 'users-urs'
    #db: Db

    async connect(connectDb: () => Promise<Db>) {
        if (!this.#db) {
            this.#db = await connectDb()
        }
    }

    async getUserIdsWithModelRoles(
        model: Model | ModelWithWorkflow,
        roles: string[]
    ) {
        if (!roles.length) {
            return []
        }

        const pipeline = [
            { $sort: { 'x-meditor.modifiedOn': -1 } }, // Sort descending by version (date)
            { $group: { _id: '$id', doc: { $first: '$$ROOT' } } }, // Grab all fields in the most recent version
            { $replaceRoot: { newRoot: '$doc' } }, // Put all fields of the most recent doc back into root of the document
            { $unwind: '$roles' },
            {
                $match: {
                    'roles.model': model.name,
                    'roles.role': { $in: roles },
                },
            },
            { $group: { _id: null, ids: { $addToSet: '$id' } } },
        ]

        const results = await this.#db
            .collection(this.#USERS_COLLECTION)
            .aggregate(pipeline, { allowDiskUse: true })
            .toArray()

        if (!results.length) {
            return []
        }

        // return only the user ids from the response
        return makeSafeObjectIDs(results[0]).ids as string[]
    }

    /**
     * given a list of mEditor users, fetches their contact information from the database
     * contact info includes: emailAddress, firstName, lastName, and uid
     */
    async getContactInformationForUsers(
        userIds: string[]
    ): Promise<UserContactInformation[]> {
        if (!userIds || !userIds.length) {
            return []
        }

        return this.#db
            .collection(this.#ACCOUNTS_COLLECTION)
            .find({ uid: { $in: userIds } })
            .project<UserContactInformation>({
                _id: 0,
                emailAddress: 1,
                firstName: 1,
                lastName: 1,
                uid: 1,
            })
            .toArray()
    }

    async createUserAccount(userContactInformation: UserContactInformation) {
        return await this.#db
            .collection<Document>(this.#ACCOUNTS_COLLECTION)
            .updateOne(
                {
                    uid: userContactInformation.uid,
                },
                {
                    $set: userContactInformation,
                },
                {
                    upsert: true,
                }
            )
    }

    async getMeditorUserByUid(uid: string): Promise<Document> {
        const user = await this.#db
            // the mEditor user record is a generic, dynamic document, like any other model's document (see the "Users" model)
            .collection<Document>(this.#USERS_COLLECTION)
            .findOne(
                {
                    id: uid,
                    'x-meditor.deletedOn': { $exists: false }, // filter out deleted users
                },
                { sort: { 'x-meditor.modifiedOn': -1 } }
            )

        return makeSafeObjectIDs(user)
    }
}

const db = new UsersDb()

async function getUsersDb() {
    await db.connect(getDb)

    return db
}

export { getUsersDb }
