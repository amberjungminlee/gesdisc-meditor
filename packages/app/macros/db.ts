import { getDb } from '../lib/connections'
import type { Db } from 'mongodb'

class MacrosDb {
    #db: Db

    async connect(connectDb: () => Promise<Db>) {
        if (!this.#db) {
            this.#db = await connectDb()
        }
    }

    async getUniqueFieldValues(fieldName: string, modelName: string) {
        const pipeline = [
            { $match: { 'x-meditor.deletedOn': { $exists: false } } }, // Do not grab deleted documents.
            { $sort: { 'x-meditor.modifiedOn': -1 } }, // Sort so that later queries can get only the latest version.
            {
                $group: {
                    _id: `$${fieldName}`,
                    field: { $first: `$${fieldName}` },
                },
            }, // Put only the first (see sort, above) desired field on the grouped document.
            { $sort: { field: 1 } }, // Sort for macro consumption.
        ]

        const documents = await this.#db
            .collection(modelName)
            .aggregate(pipeline, { allowDiskUse: true })
            .toArray()

        return documents.map(document => document.field)
    }

    async getDependenciesByTitle(dependentField: string, modelName: string) {
        const pipeline = [
            { $sort: { 'x-meditor.modifiedOn': -1 } }, // Sort descending by version (date)
            { $group: { _id: '$title', doc: { $first: '$$ROOT' } } }, // Grab all fields in the most recent version
            { $replaceRoot: { newRoot: '$doc' } }, // Put all fields of the most recent doc back into root of the document
            {
                $project: {
                    _id: 0,
                    title: 1,
                    [dependentField]: 1,
                },
            },
        ]

        const documents = await this.#db
            .collection(modelName)
            .aggregate(pipeline, { allowDiskUse: true })
            .toArray()

        return documents
    }
}

const db = new MacrosDb()

async function getMacrosDb() {
    await db.connect(getDb)

    return db
}

export { getMacrosDb }
