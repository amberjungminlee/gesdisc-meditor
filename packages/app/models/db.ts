import { getDb } from '../lib/connections'
import { makeSafeObjectIDs } from '../lib/mongodb'
import type { Db } from 'mongodb'
import type { Model } from './types'

export const MODELS_COLLECTION = 'Models'
export const MODELS_TITLE_PROPERTY = 'name'

class ModelsDb {
    #db: Db

    async connect(connectDb: () => Promise<Db>) {
        if (!this.#db) {
            this.#db = await connectDb()
        }
    }

    async getModel(modelName: string): Promise<Model> {
        const models = await this.#db
            .collection(MODELS_COLLECTION)
            .find({ [MODELS_TITLE_PROPERTY]: modelName })
            .sort({ 'x-meditor.modifiedOn': -1 })
            .limit(1)
            .toArray()

        return makeSafeObjectIDs(models)[0]
    }

    async getModels(): Promise<Model[]> {
        const models = await this.#db
            .collection(MODELS_COLLECTION)
            .aggregate(
                [
                    { $sort: { 'x-meditor.modifiedOn': -1 } }, // Sort descending by version (date)
                    { $group: { _id: '$name', doc: { $first: '$$ROOT' } } }, // Grab all fields in the most recent version
                    { $replaceRoot: { newRoot: '$doc' } }, // Put all fields of the most recent doc back into root of the document
                ],
                { allowDiskUse: true }
            )
            .toArray()

        return makeSafeObjectIDs(models)
    }
}

const db = new ModelsDb()

async function getModelsDb() {
    await db.connect(getDb)

    return db
}

export { getModelsDb }
