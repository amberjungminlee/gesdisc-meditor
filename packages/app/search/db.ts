import type { Db } from 'mongodb'
import compile from 'monquery'
import getDb, { makeSafeObjectIDs } from '../lib/mongodb'

class searchDb {
    #db: Db

    async connect(connectDb: () => Promise<Db>) {
        if (!this.#db) {
            this.#db = await connectDb()
        }
    }

    compileQuery(query: string) {
        return compile(query)
    }

    async search(
        modelName: string,
        titleProperty: string,
        query: string,
        resultsPerPage: number,
        pageNumber: number
    ): Promise<any> {
        //* The pipeline order matters: we have to sort and group by titleProperty first so that we're not matching old documents.
        const pipeline = [
            // Do not match items that have been deleted.
            {
                $match: {
                    'x-meditor.deletedOn': { $exists: false },
                },
            },
            // Sort descending by version (date).
            { $sort: { 'x-meditor.modifiedOn': -1 } },
            // Grab all fields in the most recent version.
            { $group: { _id: `$${titleProperty}`, doc: { $first: '$$ROOT' } } },
            // Put all fields of the most recent doc back into root of the document.
            { $replaceRoot: { newRoot: '$doc' } },
            // Compile Lucene syntax into MQL.
            { $match: this.compileQuery(query) },
            // Use a 1-based pageNumber for readability, but operate on a 0-based index.
            { $skip: resultsPerPage * (pageNumber - 1) },
            // Limit our results to the correct number.
            { $limit: resultsPerPage },
        ]
        const searchResults = await this.#db
            .collection(modelName)
            .aggregate(pipeline, { allowDiskUse: true })
            .toArray()

        return makeSafeObjectIDs(searchResults)
    }
}

const db = new searchDb()

async function getSearchDb() {
    await db.connect(getDb)

    return db
}

export { getSearchDb }
