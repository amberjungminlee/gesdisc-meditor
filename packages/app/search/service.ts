import compile from 'monquery'
import log from '../lib/log'
import { ErrorCode, HttpException } from '../utils/errors'

/**this searvice takes lucene query string and converts it to mongoDB query also throws an error if search query is mistyped*/

export function searchwithMonquery(searchQuery) {
    try {
        if (!searchQuery) {
            throw new HttpException(
                ErrorCode.BadRequest,
                'Search query should be typed correctly'
            )
        }

        let search = compile(searchQuery)

        return search
    } catch (error) {
        log.error(error)
        return [error, null]
    }
}
