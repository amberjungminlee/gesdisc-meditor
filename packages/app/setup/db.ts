import exampleNews from './db-seed/example-news.json'
import models from './db-seed/models.json'
import workflows from './db-seed/workflows.json'
import { getDb } from '../lib/connections'
import type { Db } from 'mongodb'
import type { UserDuringSetup } from './types'

class SetupDb {
    #EXAMPLE_NEWS_COLLECTION = 'Example News'
    #MODELS_COLLECTION = 'Models'
    #USERS_COLLECTION = 'Users'
    #WORKFLOWS_COLLECTION = 'Workflows'
    #db: Db

    async connect(connectDb: () => Promise<Db>) {
        if (!this.#db) {
            this.#db = await connectDb()
        }
    }

    async seedDb(users: UserDuringSetup[]): Promise<void> {
        const ISO_DATE = new Date().toISOString()

        const defaultRoles = ['Models', 'Workflows', 'Users', 'Example News'].map(
            model => ({
                model,
                role: 'Author',
            })
        )

        const usersWithMetadata = users.map(user => ({
            id: user.uid,
            name: user.name,
            roles: defaultRoles,
            'x-meditor': {
                model: 'Users',
                modifiedOn: ISO_DATE,
                modifiedBy: 'system',
                states: [
                    {
                        source: 'Init',
                        target: 'Draft',
                        modifiedOn: ISO_DATE,
                    },
                ],
            },
        }))

        await this.#db
            .collection(this.#USERS_COLLECTION)
            .insertMany(usersWithMetadata)
        await this.#db.collection(this.#MODELS_COLLECTION).insertMany(models)
        await this.#db.collection(this.#WORKFLOWS_COLLECTION).insertMany(workflows)
        await this.#db
            .collection(this.#EXAMPLE_NEWS_COLLECTION)
            .insertMany(exampleNews)
    }
}

const db = new SetupDb()

async function getSetupDb() {
    await db.connect(getDb)

    return db
}

export { getSetupDb }
