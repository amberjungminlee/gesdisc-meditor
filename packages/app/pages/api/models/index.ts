import type { NextApiRequest, NextApiResponse } from 'next'
import { getModels } from '../../../models/service'
import { respondAsJson } from '../../../utils/api'
import assert from 'assert'
import createError from 'http-errors'
import { withApiErrorHandler } from 'lib/with-api-error-handler'

const handler = async (req: NextApiRequest, res: NextApiResponse) => {
    assert(req.method === 'GET', new createError.MethodNotAllowed())

    const [error, models] = await getModels()

    if (error) {
        throw error
    }

    return respondAsJson(models, req, res)
}

export default withApiErrorHandler(handler)
