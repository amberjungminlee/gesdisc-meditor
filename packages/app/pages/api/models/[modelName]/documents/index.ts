import { getLoggedInUser } from 'auth/user'
import {
    patchDocumentsHeadersSchema,
    patchDocumentsInputSchema,
} from 'documents/schema'
import {
    createDocument,
    getDocumentsForModel,
    patchDocuments,
} from 'documents/service'
import type { JSONPatchDocument } from 'immutable-json-patch'
import { userCanAccessModel } from 'models/service'
import type { NextApiRequest, NextApiResponse } from 'next'
import { respondAsJson } from 'utils/api'
import {
    apiError,
    ErrorCode,
    formatZodError,
    HttpException,
    parseZodAsErrorData,
} from 'utils/errors'
import { safeParseJSON } from 'utils/json'
import type { ZodError } from 'zod'

export default async function handler(req: NextApiRequest, res: NextApiResponse) {
    const modelName = decodeURIComponent(req.query.modelName.toString())
    const user = await getLoggedInUser(req, res)

    if (!userCanAccessModel(modelName, user)) {
        return apiError(
            new HttpException(
                ErrorCode.ForbiddenError,
                'User does not have access to the requested model'
            ),
            res
        )
    }

    switch (req.method) {
        case 'GET': {
            const [error, documents] = await getDocumentsForModel(modelName, {
                ...(req.query.filter && { filter: req.query.filter.toString() }),
                ...(req.query.sort && { sort: req.query.sort.toString() }),
                ...(req.query.searchTerm && {
                    searchTerm: req.query.searchTerm.toString(),
                }),
            })

            if (error) {
                return apiError(error, res)
            }

            return respondAsJson(documents, req, res)
        }

        case 'POST': {
            if (!user) {
                return apiError(
                    new HttpException(ErrorCode.Unauthorized, 'Unauthorized'),
                    res
                )
            }

            const [parsingError, parsedDocument] = safeParseJSON(req.body)

            if (parsingError) {
                return apiError(parsingError, res)
            }

            const [documentError, data] = await createDocument(
                parsedDocument,
                modelName,
                user,
                req.query.initialState?.toString()
            )

            if (documentError) {
                return apiError(documentError, res)
            }

            const { _id, ...apiSafeDocument } = data.insertedDocument

            res.setHeader('Location', data.location)

            return respondAsJson(apiSafeDocument, req, res, {
                httpStatusCode: 201,
            })
        }

        case 'PATCH': {
            if (!user) {
                return apiError(
                    new HttpException(ErrorCode.Unauthorized, 'Unauthorized'),
                    res
                )
            }

            //* we enforce requiring the user to provide explicit identifiers for the documents to patch (we don't support all/*)
            //* the standard is to use an "If-Match" header: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/If-Match
            const [headersError, headers] = parseZodAsErrorData<any>(
                patchDocumentsHeadersSchema,
                req.headers
            )

            if (headersError) {
                return apiError(
                    formatZodError(headersError as ZodError, '`If-Match` header: '), // format the ZodError so we can use a custom error message prefix
                    res
                )
            }

            //* parse the document titles from 'If-Match'
            const documentTitles = headers['if-match']
                .split(',')
                .map(title => title.trim().replace(/^"/, '').replace(/"$/, ''))

            //* we'll also parse the list of operations the user requested, this will ensure they match the right format for the JSON patch spec
            const [parsingError, operations] = parseZodAsErrorData<JSONPatchDocument>(
                patchDocumentsInputSchema,
                req.body
            )

            if (parsingError) {
                return apiError(parsingError, res)
            }

            const [error, result] = await patchDocuments(
                documentTitles,
                modelName,
                user,
                operations
            )

            if (error) {
                return apiError(error, res)
            }

            return respondAsJson(result, req, res)
        }

        default:
            return res.status(405).end()
    }
}
