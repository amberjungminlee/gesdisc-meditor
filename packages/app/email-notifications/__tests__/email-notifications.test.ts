import type { Db } from 'mongodb'
import { getUsersDb } from '../../auth/db'
import baconUser from '../../auth/__tests__/__fixtures__/bacon-user.json'
import { getDocument } from '../../documents/service'
import alertWithPublication from '../../documents/__tests__/__fixtures__/alertWithPublication.json'
import getDb from '../../lib/mongodb'
import { getModelWithWorkflow } from '../../models/service'
import type { ModelWithWorkflow } from '../../models/types'
import howDoIFAQ from '../../models/__tests__/__fixtures__/faqs/how-do-i.json'
import alertsModel from '../../models/__tests__/__fixtures__/models/alerts.json'
import faqsModel from '../../models/__tests__/__fixtures__/models/faqs.json'
import {
    getNodesFromEdges,
    getWorkflowEdgeMatchingSourceAndTarget,
} from '../../workflows/service'
import editPublishWorkflow from '../../workflows/__tests__/__fixtures__/edit-publish.json'
import modifyReviewPublishWorkflow from '../../workflows/__tests__/__fixtures__/modify-review-publish.json'
import {
    constructEmailMessageForStateChange,
    formatUserForEmail,
    getUsersToCc,
    getUsersToNotifyOfStateChange,
    populateEmailMessageTemplate,
    shouldNotifyUsersOfStateChange,
} from '../service'

describe('Email Notifications', () => {
    let db: Db

    beforeAll(async () => {
        db = await getDb()

        // small helper function to add users to the mock database
        const addUser = async (id, name, roles) => {
            // insert mEditor user
            await db.collection('Users').insertOne({ id, name, roles })

            // insert Earthdata record
            await db.collection('users-urs').insertOne({
                uid: id,
                firstName: name,
                lastName: 'User',
                emailAddress: name.toLowerCase() + '@mock.nasa.gov',
            })
        }

        // add some fake users to test with
        await addUser('bacon', 'Bacon', [
            { model: 'Alerts', role: 'Author' },
            { model: 'FAQs', role: 'Author' },
        ])
        await addUser('eggs', 'Eggs', [
            { model: 'Alerts', role: 'Author' },
            { model: 'FAQs', role: 'Reviewer' },
        ])
        await addUser('hashbrowns', 'Hashbrowns', [
            { model: 'Alerts', role: 'Author' },
            { model: 'Alerts', role: 'Post-Publish Reviewer' },
            { model: 'FAQs', role: 'Reviewer' },
        ])
        await addUser('grits', 'Grits', [
            { model: 'Alerts', role: 'Author' },
            { model: 'Alerts', role: 'Post-Publish Reviewer' },
            { model: 'FAQs', role: 'Publisher' },
        ])
        await addUser('gravy', 'Gravy', [
            { model: 'Alerts', role: 'Author' },
            { model: 'FAQs', role: 'Publisher' },
        ])

        await db.collection('Models').insertOne(faqsModel)
        await db.collection('Models').insertOne(alertsModel)
        await db.collection('Workflows').insertOne(modifyReviewPublishWorkflow)
        await db.collection('Workflows').insertOne(editPublishWorkflow)
        await db.collection('Alerts').insertOne(alertWithPublication)
        await db.collection('FAQs').insertOne(howDoIFAQ)
    })

    afterAll(async () => {
        await db.collection('Users').deleteMany({})
        await db.collection('Models').deleteMany({})
        await db.collection('Workflows').deleteMany({})
        await db.collection('Alerts').deleteMany({})
        await db.collection('FAQs').deleteMany({})
    })

    describe('shouldNotifyUsersOfStateChange', () => {
        test.each`
            modelName | documentSourceState | documentState     | result
            ${'FAQs'} | ${'Init'}           | ${'Init'}         | ${false}
            ${'FAQs'} | ${'Init'}           | ${'Draft'}        | ${false}
            ${'FAQs'} | ${'Draft'}          | ${'Under Review'} | ${true}
            ${'FAQs'} | ${'Under Review'}   | ${'Draft'}        | ${true}
            ${'FAQs'} | ${'Under Review'}   | ${'Approved'}     | ${true}
            ${'FAQs'} | ${'Approved'}       | ${'Published'}    | ${true}
            ${'FAQs'} | ${'Approved'}       | ${'Under Review'} | ${true}
            ${'FAQs'} | ${'Published'}      | ${'Hidden'}       | ${false}
            ${'FAQs'} | ${'Hidden'}         | ${'Published'}    | ${false}
            ${'FAQs'} | ${'Draft'}          | ${'Deleted'}      | ${false}
            ${'FAQs'} | ${'Hidden'}         | ${'Deleted'}      | ${false}
            ${'FAQs'} | ${'Published'}      | ${'Deleted'}      | ${false}
        `(
            'should return `$result` for a document state of `$documentState` (source state of `$documentSourceState`) in model `$modelName`',
            async ({
                modelName,
                documentSourceState,
                documentState,
                result,
            }: {
                modelName: string
                documentSourceState: string
                documentState: string
                result: boolean
            }) => {
                const [error, model] = await getModelWithWorkflow(modelName)
                const currentEdge = model.workflow.edges.find(
                    edge =>
                        edge.source === documentSourceState &&
                        edge.target === documentState
                )

                expect(error).toBeNull()
                expect(
                    shouldNotifyUsersOfStateChange(documentState, currentEdge)
                ).toEqual(result)
            }
        )
    })

    test.each`
        documentState     | previousState     | modelName   | workflow                       | expectedUids
        ${'Init'}         | ${''}             | ${'FAQs'}   | ${modifyReviewPublishWorkflow} | ${[]}
        ${'Draft'}        | ${'Init'}         | ${'FAQs'}   | ${modifyReviewPublishWorkflow} | ${[]}
        ${'Under Review'} | ${'Draft'}        | ${'FAQs'}   | ${modifyReviewPublishWorkflow} | ${['eggs', 'hashbrowns']}
        ${'Draft'}        | ${'Under Review'} | ${'FAQs'}   | ${modifyReviewPublishWorkflow} | ${['eggs', 'hashbrowns']}
        ${'Approved'}     | ${'Under Review'} | ${'FAQs'}   | ${modifyReviewPublishWorkflow} | ${['grits', 'gravy']}
        ${'Under Review'} | ${'Approved'}     | ${'FAQs'}   | ${modifyReviewPublishWorkflow} | ${['eggs', 'hashbrowns']}
        ${'Published'}    | ${'Approved'}     | ${'FAQs'}   | ${modifyReviewPublishWorkflow} | ${['grits', 'gravy']}
        ${'Hidden'}       | ${'Published'}    | ${'FAQs'}   | ${modifyReviewPublishWorkflow} | ${[]}
        ${'Deleted'}      | ${'Draft'}        | ${'FAQs'}   | ${modifyReviewPublishWorkflow} | ${[]}
        ${'FakeState'}    | ${'Draft'}        | ${'FAQs'}   | ${modifyReviewPublishWorkflow} | ${[]}
        ${'Init'}         | ${''}             | ${'Alerts'} | ${editPublishWorkflow}         | ${[]}
        ${'Draft'}        | ${'Init'}         | ${'Alerts'} | ${editPublishWorkflow}         | ${[]}
        ${'Published'}    | ${'Draft'}        | ${'Alerts'} | ${editPublishWorkflow}         | ${['grits', 'hashbrowns']}
        ${'Hidden'}       | ${'Published'}    | ${'Alerts'} | ${editPublishWorkflow}         | ${['grits', 'hashbrowns']}
        ${'Deleted'}      | ${'Draft'}        | ${'Alerts'} | ${editPublishWorkflow}         | ${[]}
        ${'Deleted'}      | ${'Hidden'}       | ${'Alerts'} | ${editPublishWorkflow}         | ${[]}
        ${'FakeState'}    | ${'Draft'}        | ${'Alerts'} | ${editPublishWorkflow}         | ${[]}
    `(
        'getUsersToNotifyOfStateChange for $workflow.name workflow, $modelName model, and $documentState state (source $previousState), should return $expectedUids',
        async ({
            documentState,
            previousState,
            modelName,
            workflow,
            expectedUids,
        }) => {
            const [_error, model] = await getModelWithWorkflow(modelName)
            const currentEdge = previousState
                ? workflow.edges.find(
                      edge =>
                          edge.source == previousState && edge.target == documentState
                  )
                : undefined

            const users = await getUsersToNotifyOfStateChange(
                model,
                documentState,
                currentEdge
            )

            expect(users.map(user => user.uid).sort()).toEqual(expectedUids.sort())
        }
    )

    test.each`
        author     | user         | toUsers     | expectedCcs
        ${'bacon'} | ${'eggs'}    | ${[]}       | ${['bacon', 'eggs']}
        ${'bacon'} | ${'eggs'}    | ${['eggs']} | ${['bacon']}
        ${'bacon'} | ${undefined} | ${['eggs']} | ${['bacon']}
    `(
        'getUsersToCc($author, $user, $toUsers) should return CCs of $expectedCcs',
        async ({ author, user, toUsers, expectedCcs }) => {
            const ccUsers = await getUsersToCc(
                author,
                user,
                toUsers.map(user => ({ uid: user })) // turn it back into a "user" object
            )

            expect(ccUsers.map(user => user.uid).sort()).toEqual(expectedCcs.sort())
        }
    )

    test.each`
        firstName    | lastName     | emailAddress                | expectedEmail
        ${'John'}    | ${'Snow'}    | ${'johnsnow@mock.nasa.gov'} | ${'"John Snow" <johnsnow@mock.nasa.gov>'}
        ${undefined} | ${'Snow'}    | ${'johnsnow@mock.nasa.gov'} | ${'" Snow" <johnsnow@mock.nasa.gov>'}
        ${'John'}    | ${undefined} | ${'johnsnow@mock.nasa.gov'} | ${'"John " <johnsnow@mock.nasa.gov>'}
        ${'John'}    | ${'Snow'}    | ${undefined}                | ${''}
    `(
        'formatUserForEmail should return $expectedEmail for $firstName, $lastName, $emailAddress',
        async ({ firstName, lastName, emailAddress, expectedEmail }) => {
            expect(
                formatUserForEmail({
                    ...(firstName && { firstName }),
                    ...(lastName && { lastName }),
                    ...(emailAddress && { emailAddress }),
                })
            ).toEqual(expectedEmail)
        }
    )

    test.each`
        uids                             | expectedEmails
        ${undefined}                     | ${[]}
        ${[]}                            | ${[]}
        ${['bacon']}                     | ${['bacon@mock.nasa.gov']}
        ${['bacon', 'eggs']}             | ${['bacon@mock.nasa.gov', 'eggs@mock.nasa.gov']}
        ${['bacon', 'eggs', 'fakeuser']} | ${['bacon@mock.nasa.gov', 'eggs@mock.nasa.gov']}
    `(
        'getContactInformationForUsers($uids) returns expected email addresses: $expectedEmails',
        async ({ uids, expectedEmails }) => {
            const usersDb = await getUsersDb()
            const users = await usersDb.getContactInformationForUsers(uids)
            expect(users.map(user => user.emailAddress).sort()).toEqual(
                expectedEmails.sort()
            )
        }
    )

    describe('populateEmailMessageTemplate', () => {
        let model
        let targetNodes
        let currentEdge

        const defaultEmailTemplate =
            `An {{modelName}} document drafted by {{authorName}} has been marked by {{role}} {{userFirstName}} {{userLastName}} as '{{label}}' and is now in a '{{target}}' state.\n\n` +
            `An action is required to transition the document to one of the [{{targets}}] states. {{modelNotificationTemplate}}`

        beforeAll(async () => {
            const [_error, modelWithWorkflow] = await getModelWithWorkflow(
                faqsModel.name,
                'Under Review'
            )

            model = modelWithWorkflow
            targetNodes = getNodesFromEdges(model.workflow.currentEdges)
            currentEdge = getWorkflowEdgeMatchingSourceAndTarget(
                model.workflow,
                'Draft',
                'Under Review'
            )
        })

        it('should return a default email message if no custom email message is setup in the workflow', async () => {
            const emailMessage = await populateEmailMessageTemplate(
                model,
                {},
                targetNodes,
                currentEdge,
                'hashbrowns',
                baconUser,
                defaultEmailTemplate
            )

            expect(emailMessage).toMatchInlineSnapshot(`
                "An FAQs document drafted by Hashbrowns User has been marked by Author Bacon User as 'Submit for review' and is now in a 'Under Review' state.

                An action is required to transition the document to one of the [Draft, Approved] states. <p><strong>Latest Comment:</strong></p>"
            `)
        })

        it('should support a model without a notification template', async () => {
            const { notificationTemplate, ...modelWithNoTemplate } = model

            const emailMessage = await populateEmailMessageTemplate(
                modelWithNoTemplate,
                {},
                targetNodes,
                currentEdge,
                'hashbrowns',
                baconUser,
                defaultEmailTemplate
            )

            expect(emailMessage).toMatchInlineSnapshot(`
                "An FAQs document drafted by Hashbrowns User has been marked by Author Bacon User as 'Submit for review' and is now in a 'Under Review' state.

                An action is required to transition the document to one of the [Draft, Approved] states. "
            `)
        })

        it('should use the workflows configured email message instead of the default message', async () => {
            // a custom email message utilizing all of the passed in params
            const emailMessage = [
                'Model name: {{modelName}}',
                'Author name: {{ authorName }}',
                'Author UID: {{ author }}',
                'Role: {{ role }}',
                'Label: {{ label }}',
                'User First Name: {{ userFirstName }}',
                'User Last Name: {{ userLastName }}',
                'Targets: {{ targets }}',
                'Target: {{ target }}',
                'Model Notification Template: {{ modelNotificationTemplate }}',
            ].join(', ')

            // add the custom email message to the "Under Review node"
            const nodes = model.workflow.nodes.map(node => {
                return node.id === 'Under Review'
                    ? {
                          ...node,
                          emailMessage,
                      }
                    : node
            })

            // and add the nodes to the test workflowo
            const modelWithCustomEmailTemplate: ModelWithWorkflow = {
                ...model,
                workflow: {
                    ...model.workflow,
                    nodes,
                },
            }

            const emailTemplate = await populateEmailMessageTemplate(
                modelWithCustomEmailTemplate,
                {},
                targetNodes,
                currentEdge,
                'hashbrowns',
                baconUser,
                defaultEmailTemplate
            )

            expect(emailTemplate).toMatchInlineSnapshot(
                `"Model name: FAQs, Author name: Hashbrowns User, Author UID: hashbrowns, Role: Author, Label: Submit for review, User First Name: Bacon, User Last Name: User, Targets: Draft, Approved, Target: Under Review, Model Notification Template: <p><strong>Latest Comment:</strong></p>"`
            )
        })
    })

    describe('constructEmailMessageForStateChange', () => {
        let mockDate = new Date(2022, 0, 1, 0, 0, 0, 0)
        let dateSpy
        let model
        let currentEdge

        beforeAll(async () => {
            const [_error, modelWithWorkflow] = await getModelWithWorkflow(
                faqsModel.name,
                'Under Review'
            )

            model = modelWithWorkflow
            currentEdge = getWorkflowEdgeMatchingSourceAndTarget(
                model.workflow,
                'Draft',
                'Under Review'
            )
        })

        beforeEach(() => {
            dateSpy = jest
                .spyOn(global, 'Date')
                .mockImplementation(() => mockDate as unknown as string)
        })

        afterEach(() => {
            dateSpy.mockRestore()
        })

        it('should return a default email message if no custom email message is setup in the workflow', async () => {
            const [_error, faq] = await getDocument(
                howDoIFAQ.title,
                'FAQs',
                baconUser
            )
            const emailMessage = await constructEmailMessageForStateChange(
                model,
                faq,
                'Under Review',
                currentEdge,
                baconUser
            )

            expect(emailMessage).toMatchSnapshot()
        })
    })
})
