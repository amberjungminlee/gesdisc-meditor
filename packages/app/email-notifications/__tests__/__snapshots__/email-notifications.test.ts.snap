// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Email Notifications constructEmailMessageForStateChange should return a default email message if no custom email message is setup in the workflow 1`] = `
Object {
  "body": "An FAQs document drafted by Eggs User has been marked by Author Bacon User as 'Submit for review' and is now in a 'Under Review' state.

An action is required to transition the document to one of the [Draft, Approved] states. <p><strong>Latest Comment:</strong></p>",
  "cc": Array [
    "\\"Bacon User\\" <bacon@mock.nasa.gov>",
  ],
  "createdOn": "2022-01-01T00:00:00.000Z",
  "link": Object {
    "label": "How do I?",
    "url": "http://localhost/meditor/FAQs/How%20do%20I%3F",
  },
  "subject": "FAQs document is now Under Review",
  "to": Array [
    "\\"Eggs User\\" <eggs@mock.nasa.gov>",
    "\\"Hashbrowns User\\" <hashbrowns@mock.nasa.gov>",
  ],
}
`;
